### PR TITLE
rgw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error across different tenants

### PR DESCRIPTION
**Pull request by vastradparayya**
Support of same name bucket creation in different tenants with same userid

Signed-off-by: Parayya Vastrad <parayya.vastrad@seagate.com>

**Problem** - Tried to create a bucket with same name with same userid but different tenant
&emsp;tenant1
&emsp;&emsp;- user1 - create a bucket(testbucket) - success
&emsp;tenant2
&emsp;&emsp;- user1 - create a bucket(testbucket) - **ERROR: Bucket 'testbucket' already exists**

**Solution** - As bucket name are unique under a tenant. 
&emsp;Index name will be `<tenant>+"$"+<bucket_name>`
&emsp;Example - "motr.rgw.bucket.index." + `<tenant>$+info.bucket.name`;

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by sachinpunadikar — APPROVED** [review]
Changes look good
One suggestion - for more readability / clarity. We can use below way:
  string t_b_name = "";
  string t_name = info.bucket.tenant;
  if (t_name != "") {
    t_b_name = t_name+"$"+info.bucket.name;
  } else {
    t_b_name = info.bucket.name;
  }

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -552,8 +552,14 @@ int MotrUser::remove_user(const DoutPrefixProvider* dpp, optional_yield y)
 int MotrBucket::remove_bucket(const DoutPrefixProvider *dpp, bool delete_children, bool forward_to_master, req_info* req_info, optional_yield y)
 {
   int ret;
+  string tenant_bucket = info.bucket.tenant;
+  if (tenant_bucket != "") {
+    tenant_bucket += "$"+info.bucket.name;
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:557]:
```suggestion
    tenant_bucket += "$" + info.bucket.name;
```

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -685,6 +691,12 @@ int MotrBucket::put_info(const DoutPrefixProvider *dpp, bool exclusive, ceph::re
 {
   bufferlist bl;
   struct MotrBucketInfo mbinfo;
+  string tenant_bucket = info.bucket.tenant;
+  if (tenant_bucket != "") {
+    tenant_bucket += "$"+info.bucket.name;
+  } else {
+    tenant_bucket = info.bucket.name;
+  }
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:699]: This code block is repeated too many times. Better to create a function and call it with appropriate parameters.

Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:699]: also make the function inline if possible.

Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:699]: OR we can write a generic function to join with $ as well. whichever seems good. 😄 
```c++
// use
string bucket = join_strings("$", {tenant, bucket_name});

// definition
std::string join_strings(std::string connector, const std::vector<std::string>& segments) {
    string full_string = segments[0];
    for(int i = 1; i < segments.size(); i++)
        full_string += connector + segments[i];
    return full_string;
}
```

Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:699]: OR we can write a generic utility function to join strings. Whichever looks good.. 😄
```c++
// Use
string bucket = join_strings("$", {tenant, bucket});

// Definition
std::string join_strings(std::string connector, const std::vector<std::string>& segments) {
  std::string full_string = "";
  for(int i = 0; i < segments.size(); i++) {
    if(segments[i] == "")
      continue;
    if(full_string != "")
      full_string += connector;
    full_string += segments[i];
  }
  return full_string;
}
```

**Comment by sumedhak27** [comment]
Since the bucket index table is common across users and tenants, this issue is not restricted to the same user-id across tenants I think.
Error is also valid for different user-id's across tenants as well.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (jjxsg) [reply src/rgw/rgw_sal_motr.cc:699]: Great suggestion @sumedhak27 but it would be overkill in our case. Let's stick to '+' operator as of now.

**Comment by jjxsg** [comment]
> Since the bucket index table is global (not restricted across users/tenants), this issue is not restricted to the same user-id's across tenants.
Error is also valid for different user-id's across tenants as well.

@vastradparayya please check this scenario as well.

**Comment by vastradparayya** [comment]
> 

@sumedhak27 @jjxsg Yes, error is valid for different user-id's across tenants.
As bucket name are unique under a tenant

**Ceph behaviour -** 

tenant1
&emsp;user1 - create bucket(testbucket) - Success
tenant2
&emsp;user1 - create bucket(testbucket) - Success

**Motr as backend -** 

tenant1
&emsp;user1 - create bucket(testbucket) - Success
tenant2
&emsp;user1 - create bucket(testbucket) - ERROR: Bucket 'testbucket' already exists

**Comment by sumedhak27** [comment]
Hi @vastradparayya, then please remove the `same user` part from the PR name/description.
It could be something like 
`rgw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error for users across tenants` OR
`rgw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error for users of different tenants`

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -3066,7 +3203,13 @@ int MotrMultipartWriter::complete(size_t accounted_size, const std::string& etag
   char buf[32];
   snprintf(buf, sizeof(buf), "%08d", (int)part_num);
   p.append(buf);
-  string obj_part_iname = "motr.rgw.object." + head_obj->get_bucket()->get_name() + "." +
+  string tenant_bucket = head_obj->get_bucket()->get_tenant();
+  if (tenant_bucket != "") {
+    tenant_bucket += "$"+head_obj->get_bucket()->get_name();
+  } else {
+    tenant_bucket = head_obj->get_bucket()->get_name();
+  }
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:3211]: Please, create a static function for this code:

```c
static string& get_bucket_name(string &bucket, string &tenant)
{
  if (tenant != "")
    return tenant + "$" + bucket;
  else
    return bucket;
}
```
and use it everywhere.

**Comment by andriytk** [comment]
> Hi @vastradparayya, then please remove the `same user` part from the PR name/description. It could be something like `rgw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error for users across tenants` OR `rgw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error for users of different tenants`

Or, just `gw_sal_motr: [CORTX-29255] Fix BucketAlreadyExists error across different tenants`.

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (vastradparayya) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -549,11 +549,19 @@ int MotrUser::remove_user(const DoutPrefixProvider* dpp, optional_yield y)
   return 0;
 }
 
+static string get_bucket_name(string tenant, string bucket)
```
Comment (andriytk) [inline src/rgw/rgw_sal_motr.cc:552]: Why not using references in arguments as was suggested?

Comment (vastradparayya) [reply src/rgw/rgw_sal_motr.cc:552]: @andriytk it was throwing error - ` error: binding reference of type ‘std::__cxx11::string&’ {aka ‘std::__cxx11::basic_string<char>&’} to ‘const string’ {aka ‘const std::__cxx11::basic_string<char>’} discards qualifiers`

Can this be used?
```
static string get_bucket_name(std::string const& tenant, std::string const& bucket)
{
  if (tenant != "")
    return tenant + "$" + bucket;
  else
    return bucket;
}
```

**Review by sachinpunadikar — APPROVED** [review]
+1

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:552]: we can use it like
```c++
static string get_bucket_name(const std::string& tenant,  const std::string& bucket) {}
```
This is the concept of lval rval in c++.
`const &` means  the function can accept rval
`&` mean the function will accept on lval

**Review by sumedhak27 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (andriytk) [reply src/rgw/rgw_sal_motr.cc:552]: Yes, using `const` in args in fine, of course.

Comment (vastradparayya) [reply src/rgw/rgw_sal_motr.cc]: Done

**Review by andriytk — APPROVED** [review]
Good work!

**Pull request merged**